### PR TITLE
Enable servicing features without stabilizing the build in blazor-wasm

### PR DESCRIPTION
**Pull request by wtgodbe**
WIP of using servicing features without stablizing.

Right now, every build of the `blazor-wasm` branch is marked stable, which means every build publishes packages to a unique feed. The goal is to have all blazor-wasm packages published to one feed (except when we _actually_ want to build stable packages). We still need to enable servicing features, though, because we need to always build against the latest stable packages available (that is, what's in the baseline).

This PR does the following things:

- Always build against what's in the baseline for implementation assemblies (never use latestPackageReference or ProjectReference)
- Add explicit ProjectReferences between Blazor projects
- Update the baseline to 3.1.1
- Use baseline package versions for projects trying to determine PackageVersions of other projects

### CODE REVIEW

**Review by dougbu — APPROVED** [review]
Looks good but two template test failures look real e.g.
```log
error NU1603: Microsoft.AspNetCore.Components 3.1.0-ci depends on Microsoft.AspNetCore.Authorization (>= 3.1.0-ci) but Microsoft.AspNetCore.Authorization 3.1.0-ci was not found. An approximate best match of Microsoft.AspNetCore.Authorization 3.1.0-preview1.19465.3 was resolved.
```

**Comment by wtgodbe** [comment]
CC @pranavkm @javiercn @JunTaoLuo - this should let us publish all the blazor-wasm stuff to the blazor-wasm feed, once I get the tests working. Note that you'll have to manually flip the "stable" bit back to true when you want to ship stable.

**Comment by pranavkm** [comment]
I guess. I don't know what this PR means 😿

**Comment by JunTaoLuo** [comment]
Theoretically this should work but the template tests says otherwise.

**Comment by wtgodbe** [comment]
> error NU1603: Microsoft.AspNetCore.Components 3.1.0-ci depends on Microsoft.AspNetCore.Authorization (>= 3.1.0-ci) but Microsoft.AspNetCore.Authorization 3.1.0-ci was not found. An approximate best match of Microsoft.AspNetCore.Authorization 3.1.0-preview1.19465.3 was resolved.

This would be because the package for `Microsoft.AspNetCore.Authorization` isn't getting built. Will look into how these tests work when non-stable

**Comment by wtgodbe** [comment]
It's because when we're stable, the packages in the template tests depend on stable versions, which exist on nuget.org. I believe we always want that to be the case, correct @javiercn @pranavkm?

**Comment by wtgodbe** [comment]
Likely related to https://github.com/dotnet/aspnetcore/blob/05f2430bbd5e7b6929d5bae69dcdba0cb3c06e32/src/Components/Components/src/Microsoft.AspNetCore.Components.csproj#L34-L44, looking into it

**Comment by wtgodbe** [comment]
Something more sinister going on here - we're not building against baseline versions because `IsServicingBuild` isn't resolving to true. We just build against what the repo thinks it's producing, which is 3.1.0, which happens to be on nuget.org. I'll get this fixed up so that we don't produce stable versions, but do build against the baseline.

**Comment by javiercn** [comment]
@wtgodbe I'm not super sure how this works, tbh. I'm looking at you for ideas/proposals. As long as the assemblies we build reference the latest released patch for the 3.1 train, we are good.

**Comment by wtgodbe** [comment]
> I'm not super sure how this works, tbh. I'm looking at you for ideas/proposals. As long as the assemblies we build reference the latest released patch for the 3.1 train, we are good.

I have an idea for a fix here, which will involve updating the baseline files, enabling servicing features in this branch, and fixing package version resolution for the template tests. I understand today is code complete for preview1, so I'll get the fix in for preview2. Preview1 is built against 3.1.0 packages.

**Comment by wtgodbe** [comment]
@dougbu can you PTAL? Might be some more to be done here to achieve the goal, which is to always build against the baseline, except for blazor projects referencing each other.

File: src/Components/Blazor/Blazor/src/Microsoft.AspNetCore.Blazor.csproj [hunk 1]
```diff
@@ -12,6 +12,12 @@
     <Reference Include="Microsoft.Extensions.Options" />
   </ItemGroup>
 
+  <!-- References to Blazor projects are explicitly ProjectReferences, so that our reference resolution never tries to use the baseline -->
+  <ItemGroup>
+    <ProjectReference Include="..\..\Mono.WebAssembly.Interop\src\Mono.WebAssembly.Interop.csproj" />
```
Comment (dougbu) [inline src/Components/Blazor/Blazor/src/Microsoft.AspNetCore.Blazor.csproj]: Suspect this needs more metadata to avoid errors in ResolveReferences.props

**Review by dougbu — COMMENTED** [review]
Suspect you need to back up a bit on our various servicing changes
1. restore PatchConfig.props and update it to contain the packages we actually want to ship from this branch (I suggest unconditionally so the file doesn't need to change unless we add new packages)
2. set `$(IsPackageInThisPatch)` using that file (restoring old logic in Directory.Build.*)
3. change the condition at https://github.com/dotnet/aspnetcore/blob/blazor-wasm/Directory.Build.targets#L73 to use the baseline version in this branch -- despite `$(IsServicingBuild)` being `false`

File: src/Components/Blazor/Blazor/src/Microsoft.AspNetCore.Blazor.csproj (hunk 1 above)
Comment (dougbu) [reply src/Components/Blazor/Blazor/src/Microsoft.AspNetCore.Blazor.csproj]: Nevermind. The errors occur mostly for `@(PackageReference)` items

**Comment by dougbu** [comment]
Please ping when the validation builds start looking better

**Comment by wtgodbe** [comment]
> F:\workspace\_work\1\s\.dotnet\sdk\3.1.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(234,5): error NETSDK1004: Assets file 'F:\workspace\_work\1\s\artifacts\obj\Microsoft.AspNetCore.Routing\project.assets.json' not found. Run a NuGet package restore to generate this file. [F:\workspace\_work\1\s\src\Http\Routing\src\Microsoft.AspNetCore.Routing.csproj]

Could be that `References` that weren't resolved to `ProjectReferences` before, now are, and the projects in question never run Restore because they're not in ProjectsToBuild. Lookin' into it

**Comment by wtgodbe** [comment]
Looks like the projects in question were transformed to `ProjectReferences` both before and after my change, but before my change, they did have Restore run in them early enough for the build to find their assets files. That's no longer the case after my change. Looking at why.

**Comment by wtgodbe** [comment]
Before my change, calling `_GetAllRestoreProjectPathItems` on `Microsoft.AspNetCore.Components.Authorization` returns a list of 6 AspNetCore projects, while after, it returns only `Microsoft.AspNetCore.Components.Authorization` itself. Still investigating

**Comment by wtgodbe** [comment]
Ah, it's because I've moved the conversion of `Reference`s into `ProjectReferences` from a static location in `ResolveReferences.targets`, to within the `ResolveCustomReferences` target, which comes after the `Restore` target. Need to figure out a way to work around that, given our desire to use Baseline references where possible in favor of ProjectReferences.

**Comment by wtgodbe** [comment]
Superseded by https://github.com/dotnet/aspnetcore/pull/18894

**Pull request closed without merging**